crates/bindings/src/lib.rs

```diff
@@ -454,133 +325,91 @@ pub mod query {
-    /// This is exposed as `delete_by_{$field_name}` on types with `#[spacetimedb(table)]`
-    /// where the field has a unique constraint.
-    pub fn delete_by_unique_field<Table: TableType, T: FilterableValue, const COL_IDX: u16>(val: &T) -> bool {
-        let count = delete_by_field::<Table, T, COL_IDX>(val);
-        debug_assert!(count <= 1);
-        count > 0
+    impl<Tbl: Table, const COL_IDX: u16> UniqueColumnImpl<Tbl, COL_IDX>
+    where
+        Tbl::Row: FieldAccess<COL_IDX>,
+    {
+        pub fn new() -> Self {
+            Self { _marker: PhantomData }
+        }
     }
 
-    /// Updates the row of `Table`, where the column at `COL_IDX` matches `old`, to be `new` instead.
-    ///
-    /// Matching is defined by decoding to an `AlgebraicValue`
-    /// according to the column's schema and then `Ord for AlgebraicValue`.
-    ///
-    /// **NOTE:** Do not use directly.
-    /// This is exposed as `update_by_{$field_name}` on types with `#[spacetimedb(table)]`.
-    #[doc(hidden)]
-    pub fn update_by_field<Table: TableType, T: FilterableValue, const COL_IDX: u16>(old: &T, new: Table) -> bool {
-        // Delete the existing row, if any.
-        delete_by_field::<Table, T, COL_IDX>(old);
+    impl<Tbl: Table, const COL_IDX: u16> UniqueColumn for UniqueColumnImpl<Tbl, COL_IDX>
+    where
+        Tbl::Row: FieldAccess<COL_IDX>,
+    {
+        type Row = Tbl::Row;
+        type ColType = <Tbl::Row as FieldAccess<COL_IDX>>::Field;
+
+        fn find(&self, col_val: Self::ColType) -> Option<Self::Row> {
+            // Find the row with a match.
+            let mut iter = iter_by_col_eq(Tbl::table_id(), COL_IDX.into(), &col_val).unwrap();
+            with_row_buf(|buf| {
+                // We will always find either 0 or 1 rows here due to the unique constraint.
+                iter.read(buf);
+                debug_assert!(iter.is_exhausted());
+
+                if buf.is_empty() {
+                    return None;
+                }
+                let mut reader = buf.as_slice();
+                let row = bsatn::from_reader(&mut reader).unwrap();
+                assert!(
+                    reader.is_empty(),
+                    "iter_by_col_eq on unique field cannot return >1 rows"
+                );
+                Some(row)
+            })
+        }
 
-        // Insert the new row.
-        Table::insert(new);
+        fn delete(&self, col_val: Self::ColType) -> bool {
+            delete_by_col_eq(Tbl::table_id(), COL_IDX.into(), &col_val)
+                // TODO: Returning `Err` here was supposed to signify an error,
+                //       but it can also return `Err(_)` when there is nothing to delete.
+                .unwrap_or(0)
+                > 0
+        }
 
-        // TODO: For now this is always successful.
-        //       In the future, this could return what `delete_by_field` returns?
-        true
+        fn update(&self, new_row: Self::Row) -> Self::Row {
+            todo!()
```

Comment:
Impl this in terms of `self.delete` and an insert. There's not going to be a host call which implements this method.